packages/protocol/contracts/stability/Broker.sol

```diff
@@ -0,0 +1,164 @@
+pragma solidity ^0.5.13;
+pragma experimental ABIEncoderV2;
+
+import { console2 as console } from "forge-std/console2.sol";
+
+import { Ownable } from "openzeppelin-solidity/contracts/ownership/Ownable.sol";
+import { IERC20 } from "openzeppelin-solidity/contracts/token/ERC20/IERC20.sol";
+
+import { IPairManager } from "./interfaces/IPairManager.sol";
+import { IBroker } from "./interfaces/IBroker.sol";
+import { IReserve } from "./interfaces/IReserve.sol";
+import { IMentoExchange } from "./interfaces/IMentoExchange.sol";
+
+import { StableToken } from "./StableToken.sol";
+
+import { Initializable } from "../common/Initializable.sol";
+import { IStableToken } from "./interfaces/IStableToken.sol";
+
+/**
+ * @title Broker
+ * @notice The broker executes swaps and keeps track of spending limits per pair
+ */
+contract Broker is IBroker, Initializable, Ownable {
+  /* ==================== State Variables ==================== */
+
+  // Address of the broker contract.
+  IPairManager public pairManager;
+
+  // Address of the reserve.
+  IReserve public reserve;
+
+  /* ==================== Constructor ==================== */
+
+  /**
+   * @notice Sets initialized == true on implementation contracts.
+   * @param test Set to true to skip implementation initialization.
+   */
+  constructor(bool test) public Initializable(test) {}
+
+  /**
+   * @notice Allows the contract to be upgradable via the proxy.
+   * @param _pairManager The address of the PairManager contract.
+   * @param _reserve The address of the Rezerve contract.
+   */
+  function initilize(address _pairManager, address _reserve) external initializer {
+    _transferOwnership(msg.sender);
+    setPairManager(_pairManager);
+    setReserve(_reserve);
+  }
+
+  /* ==================== View Functions ==================== */
+
+  function quote(bytes32 pairId, address tokenIn, uint256 amountIn)
```

Comment:
I thought about it and my opinion is that this should be computed off-chain by clients to save gas, just like routing. And the clients. Otherwise we would pay the gas cost to transform token addresses to a pairId. I don't see the benefit of it imo. I think client applications would anyway query the pairManager for all pairs defined and populate the UI based on that and then when a swap needs to happen we have two options:

- There's only one pairId for the pair (most likely) and we use that
- There are more than one pairId for a pair and then the UI can present the different quotes or automatically choose the best

In either case I think selecting pairId should happen off-chain and the on-chain logic should be simple.

Reply:
Okay this makes sense.